### PR TITLE
Some warning removals

### PR DESCRIPTION
**Pull request by AngelEzquerra**
This PR removes many warnings and some unnecessary hints. Some of this is similar to a draft PR that @mratsim opened a few months ago, so there is some overlap with his work.

Note that one of the warnings could not really be fixed so I just disabled it. This is explained in the code and in the corresponding commit message.

### CODE REVIEW

File: src/arraymancer/laser/tensor/datatypes.nim [hunk 1]
```diff
@@ -77,13 +77,24 @@ when not defined(gcDestructors):
...
+  when (NimMajor, NimMinor, NimPatch) >= (2, 0, 0):
+    proc `=destroy`[T](storage: CpuStorageObj[T]) =
+      when T is KnownSupportsCopyMem:
+        if storage.isMemOwner and not storage.memalloc.isNil:
+          storage.memalloc.deallocShared()
+      else:
+        # The following cast removes the following warning:
+        # =destroy(storage.raw_buffer) can raise an unlisted exception: Exception
+        {.cast(raises: []).}:
+          `=destroy`(storage.raw_buffer)
```
Comment (Vindaar) [inline src/arraymancer/laser/tensor/datatypes.nim:89]: Huh, interesting.

File: src/arraymancer/tensor/operators_blas_l1.nim [hunk 2]
```diff
@@ -44,8 +43,33 @@ proc dot*[T: SomeInteger](a, b: Tensor[T]): T {.noSideEffect.} =
...
+  echo "+++++++++++2"
+  when T is K:
+    echo "T is K"
+    map2_inline(a, b, x + y)
+  elif T is float64:
+    echo "T is float64"
+    map2_inline(a, b, float64(x) + y)
+  else:
+    {.error: "Complex tensor addition only supported for tensors of the same type."}
+
```
Comment (Vindaar) [inline src/arraymancer/tensor/operators_blas_l1.nim:72]: I guess this is part of this PR by accident?

**Comment by Vindaar** [comment]
I'm pretty sure I added some of these "useless" import things (like the `KnownSupportsCopyMem` in some modules) for the doc generation. `nim doc` behaves differently for some reason (or maybe it boiled down to it attempting to 'compile' each module on its own, whereas these problems go away due to generics when imported in another module.

Given that doc generation is still partially broken, I'd merge it for now. Then do another pass soon to fix the doc generation soon. Maybe at this point we can attempt again to simplify the doc generation logic...

File: src/arraymancer/tensor/operators_blas_l1.nim (hunk 2 above)
Comment (AngelEzquerra) [reply src/arraymancer/tensor/operators_blas_l1.nim:72]: Definitely, sorry about that. I'll remove this spurious change and update the PR.

File: src/arraymancer/laser/tensor/datatypes.nim (hunk 1 above)
Comment (AngelEzquerra) [reply src/arraymancer/laser/tensor/datatypes.nim:89]: Courtesy of @beef, who I should have mentioned on the commit message :)

**Comment by Vindaar** [comment]
Yeah, the docgen is now broken again:

https://github.com/mratsim/Arraymancer/actions/runs/8387852728/job/22970876430

with:

```
Processing: /home/runner/work/Arraymancer/Arraymancer/arraymancer/docs/build/gemm.html [180/190]
stack trace: (most recent call last)
/tmp/nimblecache-946033505/nimscriptapi_4247050129.nim(210, 16)
/home/runner/work/Arraymancer/Arraymancer/arraymancer/arraymancer.nimble(283, 14) gen_docsTask
/home/runner/work/Arraymancer/Arraymancer/arraymancer/docs/docs.nim(152, 22) buildDocs
/home/runner/work/Arraymancer/Arraymancer/arraymancer/docs/docs.nim(43, 11) execAction
/home/runner/work/Arraymancer/Arraymancer/nim/lib/std/assertions.nim(41, 14) failedAssertImpl
/home/runner/work/Arraymancer/Arraymancer/nim/lib/std/assertions.nim(36, 13) raiseAssert
/home/runner/work/Arraymancer/Arraymancer/nim/lib/system/fatal.nim(53, 5) sysFatal
/home/runner/work/Arraymancer/Arraymancer/nim/lib/system/fatal.nim(53, 5) Error: unhandled exception: /home/runner/work/Arraymancer/Arraymancer/arraymancer/docs/docs.nim(43, 3) `ret == 0` Command failed: 1
cmd: cd /home/runner/work/Arraymancer/Arraymancer/arraymancer && /home/runner/work/Arraymancer/Arraymancer/nim/bin/nim doc  --hints:off --warnings:off --git.commit:master -o:/home/runner/work/Arraymancer/Arraymancer/arraymancer/docs/build/gemm.html --index:on /home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim
result:
/home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim(326, 17) template/generic instantiation of `gemm_strided` from here
/home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim(259, 36) template/generic instantiation of `dispatch` from here
/home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim(226, 14) template/generic instantiation of `apply` from here
/home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/primitives/matrix_multiplication/gemm.nim(216, 28) template/generic instantiation of `newTiles` from here
/home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/primitives/matrix_multiplication/gemm_tiling.nim(347, 43) template/generic instantiation of `align_raw_data` from here
Error: /home/runner/work/Arraymancer/Arraymancer/arraymancer/src/arraymancer/laser/private/memory.nim(9, 23) Error: undeclared identifier: 'KnownSupportsCopyMem'
candidates (edit distance, scope distance); see '--spellSuggest': 
 (11, 7): 'closureScope' [AssertionDefect]
       Tip: 21 messages have been suppressed, use --verbose to show them.
nimscriptwrapper.nim(160) execScript

    Error:  Exception raised during nimble script execution
Error: Process completed with exit code 1.
```

**Pull request merged**